packages/access-api/src/service/access-delegate.js

```diff
@@ -0,0 +1,107 @@
+import * as Server from '@ucanto/server'
+import { delegate } from '@web3-storage/capabilities/access'
+import * as Ucanto from '@ucanto/interface'
+import { createDelegationsStorage } from './delegations.js'
+
+/**
+ * access/delegate failure due to the 'with' resource not having
+ * enough storage capacity to store the delegation.
+ * https://github.com/web3-storage/specs/blob/7e662a2d9ada4e3fc22a7a68f84871bff0a5380c/w3-access.md?plain=1#L94
+ *
+ * Semantics inspired by https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/507
+ *
+ * @typedef {import('@web3-storage/capabilities/types').InsufficientStorage} InsufficientStorage
+ */
+
+/**
+ * @typedef {import('@web3-storage/capabilities/types').AccessDelegateSuccess} AccessDelegateSuccess
+ * @typedef {import('@web3-storage/capabilities/types').AccessDelegateFailure} AccessDelegateFailure
+ * @typedef {Ucanto.Result<AccessDelegateSuccess, AccessDelegateFailure>} AccessDelegateResult
+ */
+
+/**
+ * @param {object} ctx
+ * @param {import('../types/delegations').DelegationsStorage} ctx.delegations
+ * @param {HasStorageProvider} ctx.hasStorageProvider
+ */
+export function accessDelegateProvider(ctx) {
+  const handleInvocation = createAccessDelegateHandler(ctx)
+  return Server.provide(delegate, async ({ capability, invocation }) => {
+    return handleInvocation(
+      /** @type {Ucanto.Invocation<import('@web3-storage/capabilities/types').AccessDelegate>} */ (
+        invocation
+      )
+    )
+  })
+}
+
+/**
+ * @callback AccessDelegateHandler
+ * @param {Ucanto.Invocation<import('@web3-storage/capabilities/types').AccessDelegate>} invocation
+ * @returns {Promise<AccessDelegateResult>}
+ */
+
+/**
+ * @callback HasStorageProvider
+ * @param {Ucanto.DID<'key'>} did
+ * @returns {Promise<boolean>} whether the given resource has a storage provider
+ */
+
+/**
+ * @param {object} options
+ * @param {import('../types/delegations').DelegationsStorage} [options.delegations]
+ * @param {HasStorageProvider} [options.hasStorageProvider]
+ * @param {boolean} [options.allowServiceWithoutStorageProvider] - whether to allow service if the capability resource does not have a storage provider
+ * @returns {AccessDelegateHandler}
+ */
+export function createAccessDelegateHandler({
+  delegations = createDelegationsStorage(),
+  hasStorageProvider = async () => false,
+  allowServiceWithoutStorageProvider = false,
+} = {}) {
+  return async (invocation) => {
+    const capabability = invocation.capabilities[0]
+    if (
+      !allowServiceWithoutStorageProvider &&
+      !(await hasStorageProvider(capabability.with))
+    ) {
+      return {
+        name: 'InsufficientStorage',
```

Comment:
Is more `MissingProvider`...no?

Reply:
There is no notion in the protocols implemented today about a storage provider, so I want to avoid mentioning that concept and committing to it in the API indefinitely https://github.com/web3-storage/w3protocol/search?q=provider

But part of the issue here is that the [protocol spec](https://github.com/web3-storage/specs/blob/main/w3-access.md#accessdelegate) doesn't define the error types. I don't think we necessarily should add error names for all of the distinct things that could happen in our implementation here, so much as add just enough error types as are needed to make the protocol useful to a client regardless of implementation. I like `InsufficientStorage` because it's more than enough here to let the user know that retrying the request wont work, and it's also vague enough that we can use it for other situations. That was why I did it the way I did, but I could have painted the bikeshed another color. 

@alanshaw Can you document a use case for needing to distinguish `MissingProvider` from `InsufficientStorage`?

@Gozala I know sometimes you have success/failure types in mind in your specs, and sometimes even can point to earlier implementations where you specified some. Do you have any thoughts on what the `Failure` type for `access/delegate` should be?

We can also iterate on the details of this without blocking this PR, since this is all new functionality and the PR as written now disables the functionality in prod, so we can make backward-incompatible changes

Reply:
No feels the same, happy to go with `InsufficientStorage` for now.